Added DRLK-2.5 and DRLK-2.9 python script

sumaky opened the pull request:
removed existing DRLK_2_5, DRLK_2_9 yaml files as the test plan requires looping - moved to python
updated DLOG_2_1.yaml 
updated DRLK_2.4 yaml script as per test plan update
fixes https://github.com/CHIP-Specifications/chip-test-plans/issues/4515

sumaky committed:
added drlkcluster variable

sumaky committed:
Merge branch 'project-chip:master' into master

sumaky committed:
Delete src/python_testing/TC_DRLK_2_9.py

sumaky committed:
added DRLK_2.5 python script

sumaky committed:
Added Script DRLK_2.9

sumaky committed:
removed DRLK_2_5 yaml file

sumaky committed:
removed DRLK_2.9.yaml file

sumaky committed:
updated DLOG_2_1.yaml

sumaky committed:
removed redefined command clear user

sumaky committed:
removed unsed var

sumaky committed:
removed unused var

sumaky committed:
Updated TC_DRLK_2_4.yaml script as per test plan

sumaky committed:
Updated the TC to include PIN, Aliro combination and do Modify credential

restyled-commits committed:
Restyled by autopep8

sumaky committed:
updated PICS

sumaky committed:
Appended steps to check modify of both PIN and Aliro credentiaType is as expected

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kAliroEvictableEndpointKey),
+                              drlkcluster.Structs.CredentialStruct(credentialIndex=credentialIndex_1,
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kPin)]
+            await self.get_user(userIndex_1, user_name, user_unique_id1, credentiallist, Status.Success)
```
See 'brittle code' comment on step 19 above

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kAliroEvictableEndpointKey),
+                              drlkcluster.Structs.CredentialStruct(credentialIndex=credentialIndex_1,
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kPin)]
+            await self.get_user(userIndex_1, user_name, user_unique_id1, credentiallist, Status.Success)
```
See 'brittle code' comment on step 19 above

sumaky replied:
added additional check in get user command in a loop as follows:
Traverse through the input credential list and check for the match value in the response credential list

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kAliroEvictableEndpointKey),
+                              drlkcluster.Structs.CredentialStruct(credentialIndex=credentialIndex_1,
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kPin)]
+            await self.get_user(userIndex_1, user_name, user_unique_id1, credentiallist, Status.Success)
```
See 'brittle code' comment on step 19 above

sumaky replied:
added additional check in get user command in a loop as follows:
Traverse through the input credential list and check for the match value in the response credential list

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kPin)]
+            await self.get_user(userIndex_1, user_name, user_unique_id1, credentiallist, Status.Success)
+
+        self.step("35")
```
Steps 35 and 36 in different order than test plan?

sumaky replied:
corrected the order

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+
+    async def set_credential_cmd(self, credential_enum: drlkcluster.Enums.CredentialTypeEnum, statuscode, credentialIndex,
+                                 operationType, userIndex, credentialData, userStatus, userType):
+        custom_stautus_code = 149
```
Consider using a global constant for this magic number

sumaky replied:
This comment has not been addressed yet. Not sure where to update the global constant.
 Currently it is not present in init_.py Status enum. Maybe fixing init.py would solve the issue and custom_stautus_code = 149 can be removed

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            await self.set_credential_cmd(credentialData=self.pin_code,
+                                          operationType=drlkcluster.Enums.DataOperationTypeEnum.kAdd,
+                                          credential_enum=drlkcluster.Enums.CredentialTypeEnum.kPin,
+                                          credentialIndex=credentialIndex_1, userIndex=NullValue, userStatus=NullValue, userType=NullValue, statuscode=149)
```
Consider using a global constant for this magic number

sumaky replied:
This comment has not been addressed yet. Not sure where to update the global constant.
Currently it is not present in init_.py Status enum. Maybe fixing init.py would solve the issue and custom_stautus_code = 149 can be removed

sumaky committed:
Updated DRLK-2.5 test case as per PR comment
1.correted test step description for step 4,5,6 and 9
2.Corrected steps 5,6 and 9

sumaky committed:
Updated DRLK-2.9
1. Corrected test step description
2. Corrected PICS
3. Validated arrtibutes read based on respective range
4. modified get_user function to check for given credentialsirespective to the order of credentials listed

sumaky commented on src/python_testing/TC_DRLK_2_5.py:
```diff
@@ -0,0 +1,390 @@
+            TestStep("3", "TH sends GetWeekDaySchedule Command to DUT.",
+                     "Verify that the DUT sends GetWeekDayScheduleResponse command with expected values."),
+            TestStep("4", "TH sends SetWeekDaySchedule Command to DUT.",
+                     "Verify that the DUT responds with SetCredentialResponse command with Status INVALID_COMMAND."),
```
Corrected

sumaky commented on src/python_testing/TC_DRLK_2_5.py:
```diff
@@ -0,0 +1,390 @@
+                     "Verify that the DUT responds with SetCredentialResponse command with Status INVALID_COMMAND."),
+            TestStep("8", "TH sends GetWeekDaySchedule Command to DUT.",
+                     "Verify that the DUT responds with SetCredentialResponse command with Status INVALID_COMMAND."),
+            TestStep("9", "TH sends GetWeekDaySchedule Command to DUT.",
```
corrected the expected Status

sumaky commented on src/python_testing/TC_DRLK_2_5.py:
```diff
@@ -0,0 +1,390 @@
+            number_week_day_schedules_supported_per_user = await self.read_attributes_from_dut(endpoint=self.app_cluster_endpoint,
+                                                                                               cluster=drlkcluster,
+                                                                                               attribute=Clusters.DoorLock.Attributes.NumberOfWeekDaySchedulesSupportedPerUser)
+            logging.info("NumberOfWeekDaySchedulesSupportedPerUser %s" % (number_week_day_schedules_supported_per_user))
```
Added the attribute range check.

sumaky commented on src/python_testing/TC_DRLK_2_5.py:
```diff
@@ -0,0 +1,390 @@
+                Status.InvalidCommand)
+        self.step("5")
+        if self.pics_guard(self.check_pics("DRLK.S.F04") and self.check_pics("DRLK.S.C0b.Rsp")):
+            week_day_index = 0  # invalid value
```
Corrected the step

sumaky commented on src/python_testing/TC_DRLK_2_5.py:
```diff
@@ -0,0 +1,390 @@
+                Status.InvalidCommand)
+        self.step("6")
+        if self.pics_guard(self.check_pics("DRLK.S.F04") and self.check_pics("DRLK.S.C0b.Rsp")):
+            week_day_index = 1  # valid value
```
Modified Step 5 with WeekDayIndex=1, DayMaskMap=0
Modified step 6 with WeekDayIndex=1, DayMaskMap=128

sumaky commented on src/python_testing/TC_DRLK_2_5.py:
```diff
@@ -0,0 +1,390 @@
+                start_Minute,
+                end_Hour,
+                end_Minute,
+                Status.InvalidCommand)
```
corrected the Status

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+
+            TestStep("1", "TH reads NumberOfTotalUsersSupported attribute.",
+                     "Verify that TH is able to read the attribute successfully."),
+            TestStep("2a", "TH sends SetUser Command to DUT.", "Verify that TH is able to read the attribute successfully."),
```
corrected the description

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            TestStep("2d", "TH reads MinRFIDCodeLength attribute.",
+                     "Verify that TH is able to read the attribute successfully."),
+            TestStep("2e", "TH reads MaxRFIDCodeLength attribute.",
+                     "Verify that TH is able to read the attribute successfully."),
```
corrected the description

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            TestStep("14c", "TH sends ClearUser Command to DUT to clear all the users.",
+                     "Verify that the DUT sends SUCCESS response."),
+            TestStep("15a", "TH reads NumberOfCredentialsSupportedPerUser attribute from DUT.",
+                     "Verify that TH is able to read the attribute successfully."),
```
corrected the description

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                     "Verify that the DUT sends SetCredentialResponse command with Status RESOURCE_EXHAUSTION."),
+            TestStep("15e", "TH sends ClearCredential Command to DUT to clear all the credentials of PIN type.",
+                     "Verify that the DUT sends SUCCESS response."),
+            TestStep("16", " TH sends SetUser Command to DUT.",
```
yes missed step 15f. Now added as Clear all users

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                     "Verify that the DUT responds with SetCredentialResponse command with status success "),
+            TestStep("34", "TH sends GetUser Command to DUT.",
+                     "Verify that the DUT sends SUCCESS response."),
+            TestStep("35", "TH sends ClearCredential Command to DUT.",
```
corrected the order

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                                          credential_enum=drlkcluster.Enums.CredentialTypeEnum.kRfid,
+                                          credentialIndex=credentialIndex_1, userIndex=userIndex_1, userStatus=NullValue, userType=NullValue, statuscode=Status.Success)
+        self.step("18")
+        if self.pics_guard(self.check_pics("DRLK.S.F00") and self.check_pics("DRLK.S.F08")
```
corrected the PICS- removed DRLK.S.F08

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            await self.get_user(userIndex_1, user_name, user_unique_id1, credentiallist, Status.Success)
+
+        self.step("20")
+        if self.pics_guard(self.check_pics("DRLK.S.F00") and self.check_pics("DRLK.S.F08")
```
corrected the PICS- removed DRLK.S.F08

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            await self.get_user(userIndex_1, user_name, user_unique_id1, credentiallist, Status.Success)
+
+        self.step("22")
+        if self.pics_guard(self.check_pics("DRLK.S.C26.Rsp")):
```
corrected the PICS- added DRLK.S.F08

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            try:
+                await self.send_single_cmd(cmd=drlkcluster.Commands.SetUser(
+                    operationType=Clusters.DoorLock.Enums.DataOperationTypeEnum.kAdd,
+                    userIndex=1,
```
set userIndex to userIndex_1

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            except InteractionModelError as e:
+                logging.exception(e)
+
+        self.step("2b")
```
added the range check

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            await self.set_credential_cmd(credentialData=self.inavlid_pincode,
+                                          operationType=drlkcluster.Enums.DataOperationTypeEnum.kAdd,
+                                          credential_enum=drlkcluster.Enums.CredentialTypeEnum.kPin,
+                                          credentialIndex=credentialIndex_2, userIndex=userIndex_1, userStatus=invalid_user_status, userType=invalid_user_type, statuscode=Status.InvalidCommand)
```
corrected userIndex as NullValue

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                    operationType=Clusters.DoorLock.Enums.DataOperationTypeEnum.kAdd,
+                    userIndex=userIndex_2,
+                    userName=user_name,
+                    userUniqueID=user_unique_id2,
```
Used user_unique_id. A small correction is required in the test plan. 
followup issue:
https://github.com/CHIP-Specifications/chip-test-plans/issues/4684

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+
+        self.step("14c")
+        if self.pics_guard(self.check_pics("DRLK.S.F08") and self.check_pics("DRLK.S.C1d.Rsp")):
+            await self.send_clear_user_cmd(userIndex_1)
```
corrected to clear all the users using 0xFFFE

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+
+            numberofcredentialsupportedperuser = await self.read_attributes_from_dut(endpoint=self.app_cluster_endpoint,
+                                                                                     cluster=drlkcluster,
+                                                                                     attribute=Clusters.DoorLock.Attributes.NumberOfCredentialsSupportedPerUser)
```
Validated the range

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                                                  operationType=drlkcluster.Enums.DataOperationTypeEnum.kAdd,
+                                                  credential_enum=drlkcluster.Enums.CredentialTypeEnum.kPin,
+                                                  credentialIndex=start_credential_index, userIndex=userIndex_1, userStatus=NullValue, userType=NullValue, statuscode=Status.Success)
+                    start_credential_index += 1
```
validated the nextCredentialIndex

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            except InteractionModelError as e:
+                logging.exception(e)
+
+        self.step("15c")
```
Added the above check

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+            credentials = drlkcluster.Structs.CredentialStruct(credentialIndex=0xFFFE,
+                                                               credentialType=drlkcluster.Enums.CredentialTypeEnum.kPin)
+            await self.clear_credentials_cmd(credential=credentials)
+
```
Added Clear all users step

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kRfid),
+                              drlkcluster.Structs.CredentialStruct(credentialIndex=credentialIndex_1,
+                                                                   credentialType=drlkcluster.Enums.CredentialTypeEnum.kPin)]
+            await self.get_user(userIndex_1, user_name, user_unique_id1, credentiallist, Status.Success)
```
added additional check in get user command in a loop as follows:
Traverse through the input credential list and check for the match value in the response credential list

sumaky committed:
Added nextcredentialIndex check in get_credential_status function

sumaky committed:
fixed code-lints error

sumaky committed:
Added generate_max_pincode_len for generating valid pin code of Max pincode length

restyled-commits committed:
Restyled by isort

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -202,15 +209,21 @@ async def get_user(self, userindex, username, useruniqueid, credentiallist, expe
+            # traverse  through input credentials and match each value with the resonse credential
+            for input_credential_index in range(len(credentiallist)):
+                match_found = False
+                for response_credential_index in range(len(response.credentials)):
```
While this loop now successfully allows credentials in the response to be in different order, it does not fully verify the test case: Need to add a check that the length of credentialList is equal to the length of response.credentials (to detect if response.credentials possibly has extra unexpected entries)

sumaky replied:
Added the check though test plan does not mention this check. Will update the issue with this check

jrhees-cae replied:
Hmm, interesting point. Since user is created by this TC, there should not be any unexpected/additional credentials.

sumaky committed:
added a check in get_user funtion that length of given Credential list matches with length of response credential list

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,793 @@
+        self.pin_code2 = b""
+        self.inavlid_pincode = b""
+        self.rfid_tag = b""
+
```
Need assignments for maxpincodelength and maxrfidcodelength, eg: "self.maxpincodelength = None, self.maxrfidcodelength = None"

sumaky replied:
Added = None assignment for the rid and pin variables

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,793 @@
+                                                              cluster=drlkcluster,
+                                                              attribute=Clusters.DoorLock.Attributes.FeatureMap)
+            aliro_enabled = feature_map & Clusters.DoorLock.Bitmaps.Feature.kAliroProvisioning
+        if (aliro_enabled):
```
This conditional needs further indent, I believe

sumaky replied:
corrected the indent

jrhees-cae commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,793 @@
+                                                              attribute=Clusters.DoorLock.Attributes.FeatureMap)
+            aliro_enabled = feature_map & Clusters.DoorLock.Bitmaps.Feature.kAliroProvisioning
+        if (aliro_enabled):
+            credentials = drlkcluster.Structs.CredentialStruct(credentialIndex=credentialIndex_1,
```
Assignment of 'credentials' needs to be moved before the 'if (aliro_enabled)' line. Otherwise, the 'clearCredential' command in the else clause uses credentials list from step 12

sumaky replied:
corrected

sumaky committed:
PR review -Init the pin and rid variables

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,793 @@
+        self.pin_code2 = bytes(validpincodestr, 'ascii')
+        inavlidpincodestr = await self.generate_max_pincode_len(self.maxpincodelength+1)
+        self.inavlid_pincode = bytes(inavlidpincodestr, 'ascii')
+        validpincodestr = await self.generate_max_pincode_len(self.maxrfidcodelength)
```
Added the not None check

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,793 @@
+
+    async def generate_code(self):
+
+        validpincodestr = await self.generate_max_pincode_len(self.maxpincodelength)
```
Added the maxpincodelength != None check

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,793 @@
+            credentialType=credential_enum,
+            credentialIndex=credentialIndex)
+        try:
+            logging.info("maxpincodelength value is %s" % (self.maxpincodelength))
```
removed the logging statement

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,793 @@
+            credentialIndex=credentialIndex)
+        try:
+            logging.info("maxpincodelength value is %s" % (self.maxpincodelength))
+            logging.info("maxrfidcodelength value is %s" % (self.maxrfidcodelength))
```
removed logging statement

sumaky committed:
fixed lint error

sumaky commented on src/python_testing/TC_DRLK_2_9.py:
```diff
@@ -0,0 +1,742 @@
+        self.step("1")
+        if self.pics_guard(self.check_pics("DRLK.S.F08") and self.check_pics("DRLK.S.A0011")):
+            self.numberoftotaluserssupported = await self.read_attributes_from_dut(endpoint=self.app_cluster_endpoint,
+                                                                                   cluster=drlkcluster,
```
Added the validation for the range
